File: src/main/python/com/example/tutorial/dsa/medium/concurrency/counter/lock_counter.py
"""
LockCounter - Thread-safe counter using threading.Lock.

Uses a Lock to ensure only one thread can modify the counter at a time.
Equivalent to Java's ReentrantLock.

How it works:
    Thread A acquires lock
    Thread A increments count (5 -> 6)
    Thread A releases lock
    Thread B acquires lock
    Thread B increments count (6 -> 7)
    Thread B releases lock

Time Complexity: O(1) for increment and get
Space Complexity: O(1)
"""

import threading


class LockCounter:
    """Thread-safe counter using Lock."""

    def __init__(self):
        self._count = 0
        self._lock = threading.Lock()

    def increment(self) -> None:
        """Increments counter with lock protection."""
        with self._lock:
            self._count += 1

    def get_count(self) -> int:
        """Returns current count with lock protection."""
        with self._lock:
            return self._count

File: src/main/java/com/example/tutorial/dsa/medium/concurrency/counter/LockCounter.java
package com.example.tutorial.dsa.medium.concurrency.counter;

import java.util.concurrent.locks.ReentrantLock;

/**
 * Counter using ReentrantLock for explicit locking.
 *
 * <p><b>LOGIC:</b> Explicit lock/unlock calls provide more control
 * than synchronized blocks.
 *
 * <p><b>How it works:</b>
 * <ol>
 *   <li>Acquire lock explicitly with lock.lock()</li>
 *   <li>Perform operation in try block</li>
 *   <li>Release lock in finally block (guaranteed!)</li>
 * </ol>
 *
 * <p><b>Time Complexity: O(1)</b> for increment (excluding wait time).
 *
 * <p><b>Advantages over synchronized:</b>
 * <ul>
 *   <li>tryLock() - attempt lock without blocking</li>
 *   <li>tryLock(timeout) - wait with timeout</li>
 *   <li>lockInterruptibly() - can be interrupted while waiting</li>
 *   <li>Fair locking option - prevents starvation</li>
 *   <li>Multiple condition variables</li>
 * </ul>
 *
 * <p><b>Important:</b> Always use try-finally to ensure unlock!
 * <pre>
 * lock.lock();
 * try {
 *   // critical section
 * } finally {
 *   lock.unlock();  // ALWAYS in finally!
 * }
 * </pre>
 *
 * <p><b>When to use:</b> When you need more control than synchronized provides,
 * such as tryLock, timeouts, or fair locking.
 */
public class LockCounter implements Counter {
  private int count = 0;
  private final ReentrantLock lock = new ReentrantLock();

  @Override
  public void increment() {
    lock.lock();
    try {
      count++;
    } finally {
      lock.unlock();
    }
  }

  @Override
  public int get() {
    lock.lock();
    try {
      return count;
    } finally {
      lock.unlock();
    }
  }

  /**
   * Attempts to increment without blocking.
   *
   * @return true if increment succeeded, false if lock was unavailable
   */
  public boolean tryIncrement() {
    if (lock.tryLock()) {
      try {
        count++;
        return true;
      } finally {
        lock.unlock();
      }
    }
    return false;
  }
}
